File: pet-store-frontend/src/app/services/pet.service.ts
import { Injectable } from "@angular/core";
import { HttpParams } from "@angular/common/http";
import { Observable } from "rxjs";
import { Pet, PetStatus } from "../models/pet.model";

export interface PetPageResponse {
  pets: Pet[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
}
import { BaseApiService } from "./base-api.service";

@Injectable({
  providedIn: "root",
})
export class PetService extends BaseApiService {
  private readonly endpoint = "pets";

  /**
   * Fetches all pets with pagination and optional filters.
   * GET /pets?page={page}&size={size}&name={name}&categoryId={categoryId}&status={status}
   */
  getAllPets(
    page: number = 0,
    size: number = 10,
    filters?: { name?: string; categoryId?: number; status?: PetStatus }
  ): Observable<PetPageResponse> {
    let params = new HttpParams()
      .set("page", page.toString())
      .set("size", size.toString());
    if (filters?.name) params = params.set("name", filters.name);
    if (filters?.categoryId)
      params = params.set("categoryId", filters.categoryId.toString());
    if (filters?.status) params = params.set("status", filters.status);
    return this.http.get<PetPageResponse>(this.getApiUrl(this.endpoint), {
      params,
    });
  }

  getLatestPets(limit?: number): Observable<Pet[]> {
    const params = limit ? new HttpParams().set("limit", limit.toString()) : undefined;
    return this.http.get<Pet[]>(this.getApiUrl(`${this.endpoint}/latest`), { params });
  }

  /**
   * Fetches a single pet by its ID.
   * GET /pets/{id}
   */
  getPetById(id: number): Observable<Pet> {
    return this.http.get<Pet>(this.getApiUrl(`${this.endpoint}/${id}`));
  }

  /**
   * Searches pets by name, category, or status.
   * GET /pets?name={name}&categoryId={categoryId}&status={status}
   */
  searchPets(filters: {
    name?: string;
    categoryId?: number;
    status?: PetStatus;
  }): Observable<Pet[]> {
    let params = new HttpParams();

    if (filters.name) {
      params = params.set("name", filters.name);
    }
    if (filters.categoryId) {
      params = params.set("categoryId", filters.categoryId.toString());
    }
    if (filters.status) {
      params = params.set("status", filters.status);
    }

    return this.http.get<Pet[]>(this.getApiUrl(this.endpoint), { params });
  }

  /**
   * Adds a new pet.
   * POST /pets
   */
  addPet(pet: Pet): Observable<Pet> {
    return this.http.post<Pet>(this.getApiUrl(this.endpoint), pet);
  }

  /**
   * Updates an existing pet by its ID.
   * PUT /pets/{id}
   */
  updatePet(id: number, pet: Pet): Observable<Pet> {
    return this.http.put<Pet>(this.getApiUrl(`${this.endpoint}/${id}`), pet);
  }

  /**
   * Deletes a pet by its ID.
   * DELETE /pets/{id}
   */
  deletePet(id: number): Observable<void> {
    return this.http.delete<void>(this.getApiUrl(`${this.endpoint}/${id}`));
  }

  /**
   * Updates the status of a pet by its ID.
   * POST /pets/{id}/status?status={status}
   */
  updatePetStatus(id: number, status: PetStatus): Observable<Pet> {
    const params = new HttpParams().set("status", status);
    return this.http.post<Pet>(
      this.getApiUrl(`${this.endpoint}/${id}/status`),
      null,
      { params }
    );
  }

  /**
   * Fetches pets owned by the current user with pagination and optional filters.
   * GET /pets/my-pets?page={page}&size={size}&name={name}&categoryId={categoryId}&status={status}
   */
  getMyPets(
    page: number = 0,
    size: number = 10,
    filters?: { name?: string; categoryId?: number; status?: PetStatus }
  ): Observable<PetPageResponse> {
    let params = new HttpParams()
      .set("page", page.toString())
      .set("size", size.toString());
    if (filters?.name) params = params.set("name", filters.name);
    if (filters?.categoryId)
      params = params.set("categoryId", filters.categoryId.toString());
    if (filters?.status) params = params.set("status", filters.status);
    return this.http.get<PetPageResponse>(this.getApiUrl(`${this.endpoint}/my-pets`), {
      params,
    });
  }
}

File: pet-store-api/src/main/java/com/petstore/service/PetService.java
package com.petstore.service;

import com.petstore.model.Pet;
import com.petstore.enums.PetStatus;
import com.petstore.exception.InvalidPetException;
import com.petstore.exception.PetNotFoundException;
import com.petstore.exception.CategoryNotFoundException;
import com.petstore.model.Category;
import com.petstore.repository.PetRepository;
import com.petstore.repository.CategoryRepository;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Service for managing pets in the store
 */
@Service
public class PetService {

    private final PetRepository petRepository;

    private final CategoryRepository categoryRepository;

    public PetService(PetRepository petRepository, CategoryRepository categoryRepository) {
        this.petRepository = petRepository;
        this.categoryRepository = categoryRepository;
    }

    /**
     * Searches pets using multiple filters with pagination
     *
     * @param name       optional pet name filter
     * @param categoryId optional category filter
     * @param status     optional status filter
     * @param page       page number (zero-based)
     * @param size       page size
     * @return paginated result of pets
     */
    public Page<Pet> findPetsByFiltersPaginated(String name, Long categoryId, PetStatus status, Long userId, int page,
            int size) {
        Pageable pageable = PageRequest.of(page, size);
        return petRepository.findPetsByFiltersPaginated(name, categoryId, status, userId, pageable);
    }

    /**
     * Retrieves a pet by its ID
     *
     * @param id the pet ID
     * @return the pet if found
     */
    public Optional<Pet> getPetById(Long id) {
        return petRepository.findById(id);
    }

    /**
     * Gets the most recently added available pets
     *
     * @param limit maximum number of pets to return
     * @return list of available pets ordered by creation date
     */
    public List<Pet> getLatestAvailablePets(int limit) {
        Pageable pageable = PageRequest.of(0, limit);
        return petRepository.findLatestPetsByStatus(PetStatus.AVAILABLE, pageable);
    }

    /**
     * Creates a new pet. Returns null if input is null.
     *
     * @param pet the pet details to save
     * @return the created pet, or null if input is null
     */
    public Pet savePet(Pet pet) {
        
        if (pet == null) {
            throw new InvalidPetException("Pet cannot be null");
        }
        return petRepository.save(pet);
    }

    /**
     * Updates an existing pet
     *
     * @param id         the pet ID to update
     * @param petDetails the new pet details
     * @return the updated pet, or null if not found
     */
    public Pet updatePet(Long id, Pet petDetails) {

        Pet existingPet = petRepository.findById(id)
                .orElseThrow(() -> new PetNotFoundException(id));

        // Update allowed fields
        existingPet.setName(petDetails.getName());
        existingPet.setDescription(petDetails.getDescription());
        existingPet.setPrice(petDetails.getPrice());
        existingPet.setStatus(petDetails.getStatus());
        existingPet.setPhotoUrls(petDetails.getPhotoUrls());
        existingPet.setTags(petDetails.getTags());

        // Validate and set category
        if (petDetails.getCategory() != null && petDetails.getCategory().getId() != null) {
            Category category = categoryRepository.findById(petDetails.getCategory().getId())
                    .orElseThrow(() -> new CategoryNotFoundException(petDetails.getCategory().getId()));
            existingPet.setCategory(category);
        }

        // Optional: only admins can reassign ownership
        if (petDetails.getOwner() != null) {
            existingPet.setOwner(petDetails.getOwner());
        }

        return petRepository.save(existingPet);
    }

    /**
     * Deletes a pet
     *
     * @param id the ID of the pet to delete
     * @return true if deleted, false if not found
     */
    public Boolean deletePet(Long id) {

        if (id == null) {
            throw new InvalidPetException("Pet ID cannot be null");
        }

        Pet pet = petRepository.findById(id)
                .orElseThrow(() -> new PetNotFoundException(id));

        petRepository.delete(pet);
        return true;
    }

    /**
     * Updates a pet's status
     *
     * @param id     the pet ID to update
     * @param status the new status to set
     * @return the updated pet, or null if not found
     */
    public Pet updatePetStatus(Long id, PetStatus status) {
        Optional<Pet> existingPet = petRepository.findById(id);

        if (existingPet.isPresent()) {
            if (status == null) {
                throw new InvalidPetException("Pet status cannot be null");
            }
            Pet pet = existingPet.get();
            pet.setStatus(status);
            return petRepository.save(pet);
        }

        return null;
    }
}